File: android/app/src/main/java/com/HammersTech/RoutineChart/features/parent/dashboard/ParentDashboardViewModel.kt
package com.HammersTech.RoutineChart.features.parent.dashboard

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.HammersTech.RoutineChart.core.domain.models.ChildProfile
import com.HammersTech.RoutineChart.core.domain.models.Routine
import com.HammersTech.RoutineChart.core.data.remote.firebase.CompositeRoutineRepository
import com.HammersTech.RoutineChart.core.domain.repositories.AuthRepository
import com.HammersTech.RoutineChart.core.domain.repositories.ChildProfileRepository
import com.HammersTech.RoutineChart.core.domain.repositories.FamilyRepository
import com.HammersTech.RoutineChart.core.domain.repositories.RoutineRepository
import com.HammersTech.RoutineChart.core.domain.repositories.UserRepository
import com.HammersTech.RoutineChart.core.utils.AppLogger
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * ViewModel for parent dashboard
 */
@HiltViewModel
class ParentDashboardViewModel @Inject constructor(
    private val familyRepository: FamilyRepository,
    private val routineRepository: RoutineRepository,
    private val childProfileRepository: ChildProfileRepository,
    private val authRepository: AuthRepository,
    private val userRepository: UserRepository
) : ViewModel() {

    private val _state = MutableStateFlow(ParentDashboardState())
    val state: StateFlow<ParentDashboardState> = _state.asStateFlow()

    /** Tracks which user's data is currently loaded so we reload when the user switches */
    private var lastLoadedUserId: String? = null

    init {
        viewModelScope.launch {
            authRepository.authStateFlow.collect { authUser ->
                val newId = authUser?.id
                if (newId != lastLoadedUserId) {
                    lastLoadedUserId = newId
                    if (newId == null) {
                        _state.update {
                            it.copy(
                                routines = emptyList(),
                                children = emptyList(),
                                familyId = null,
                                isLoading = false,
                                error = "Not signed in"
                            )
                        }
                    } else {
                        _state.update {
                            it.copy(
                                routines = emptyList(),
                                children = emptyList(),
                                familyId = null,
                                isLoading = true,
                                error = null
                            )
                        }
                        loadData()
                    }
                }
            }
        }
    }

    fun loadData() {
        viewModelScope.launch {
            try {
                _state.update { it.copy(isLoading = true, error = null) }

                // Get current authenticated user to find their familyId
                val authUser = authRepository.currentUser
                if (authUser == null) {
                    _state.update { it.copy(isLoading = false, error = "Not signed in") }
                    return@launch
                }
                
                // Get user record to find their familyId
                val user = userRepository.getById(authUser.id)
                if (user == null) {
                    _state.update { it.copy(isLoading = false, error = "User not found. Please join a family first.") }
                    return@launch
                }

                // CRITICAL: Pull routines from Firestore BEFORE loading from local
                // This ensures we have the latest data when the UI loads
                if (routineRepository is CompositeRoutineRepository) {
                    try {
                        // First, upload any unsynced local changes
                        val uploaded = routineRepository.uploadUnsynced(user.id, user.familyId)
                        if (uploaded > 0) {
                            AppLogger.UI.info("✅ Uploaded $uploaded unsynced routine(s) before loading dashboard")
                        }
                        
                        // Then, pull any remote changes (this will also pull steps)
                        val pulled = routineRepository.pullRoutines(user.id, user.familyId)
                        if (pulled > 0) {
                            AppLogger.UI.info("✅ Pulled $pulled routine(s) from Firestore before loading dashboard")
                        }
                    } catch (e: Exception) {
                        AppLogger.UI.error("⚠️ Failed to sync routines before loading dashboard: ${e.message}", e)
                        // Continue loading local data even if sync fails
                    }
                }

                // Load routines (exclude deleted) for the user (by userId, optionally filtered by familyId)
                val allRoutines = routineRepository.getAll(user.id, user.familyId, includeDeleted = false)
                val activeRoutines = allRoutines.filter { it.deletedAt == null }
                    .sortedBy { it.createdAt }

                // Load children for the user's family
                val children = childProfileRepository.getByFamilyId(user.familyId)

                _state.update {
                    it.copy(
                        familyId = user.familyId,
                        routines = activeRoutines,
                        children = children,
                        isLoading = false
                    )
                }

                AppLogger.UI.info("Loaded ${activeRoutines.size} routines and ${children.size} children for family: ${user.familyId}")
            } catch (e: Exception) {
                AppLogger.UI.error("Error loading dashboard data: ${e.message}", e)
                _state.update { it.copy(isLoading = false, error = e.message) }
            }
        }
    }

    fun deleteRoutine(routineId: String) {
        viewModelScope.launch {
            try {
                // Soft delete by setting deletedAt
                val routine = routineRepository.getById(routineId)
                if (routine != null) {
                    val deleted = routine.copy(deletedAt = java.time.Instant.now())
                    routineRepository.update(deleted)
                    loadData() // Reload
                    AppLogger.UI.info("Deleted routine: $routineId")
                }
            } catch (e: Exception) {
                AppLogger.UI.error("Error deleting routine: ${e.message}", e)
                _state.update { it.copy(error = "Failed to delete routine") }
            }
        }
    }
    
    fun signOut() {
        viewModelScope.launch {
            try {
                authRepository.signOut()
                AppLogger.UI.info("User signed out")
            } catch (e: Exception) {
                AppLogger.UI.error("Error signing out: ${e.message}", e)
                _state.update { it.copy(error = "Failed to sign out") }
            }
        }
    }
}

data class ParentDashboardState(
    val isLoading: Boolean = true,
    val error: String? = null,
    val familyId: String? = null,
    val routines: List<Routine> = emptyList(),
    val children: List<ChildProfile> = emptyList()
)

File: ios/RoutineChart/RoutineChart/Features/Parent/Dashboard/ParentDashboardViewModel.swift
import Foundation
import Combine
import OSLog

@MainActor
final class ParentDashboardViewModel: ObservableObject {
    @Published var routines: [Routine] = []
    @Published var children: [ChildProfile] = []
    @Published var isLoading = false
    @Published var errorMessage: String?
    
    private let routineRepository: RoutineRepository
    private let childProfileRepository: ChildProfileRepository
    private let familyRepository: FamilyRepository
    private let authRepository: AuthRepository
    private let userRepository: UserRepository
    
    private var familyId: String?
    
    init(
        routineRepository: RoutineRepository,
        childProfileRepository: ChildProfileRepository,
        familyRepository: FamilyRepository,
        authRepository: AuthRepository,
        userRepository: UserRepository
    ) {
        self.routineRepository = routineRepository
        self.childProfileRepository = childProfileRepository
        self.familyRepository = familyRepository
        self.authRepository = authRepository
        self.userRepository = userRepository
    }
    
    func loadData() async {
        isLoading = true
        errorMessage = nil
        
        do {
            // Get current authenticated user to find their familyId
            guard let authUser = authRepository.currentUser else {
                errorMessage = "Not signed in"
                isLoading = false
                return
            }
            
            // Get user record to find their familyId
            guard let user = try await userRepository.get(id: authUser.id) else {
                errorMessage = "User not found. Please join a family first."
                isLoading = false
                return
            }
            
            familyId = user.familyId
            
            // CRITICAL: Pull routines from Firestore BEFORE loading from local
            // This ensures we have the latest data when the UI loads
            if let compositeRepo = routineRepository as? CompositeRoutineRepository {
                do {
                    // First, upload any unsynced local changes
                    let uploaded = try await compositeRepo.uploadUnsynced(familyId: user.familyId)
                    if uploaded > 0 {
                        AppLogger.ui.info("✅ Uploaded \(uploaded) unsynced routine(s) before loading dashboard")
                    }
                    
                    // Then, pull any remote changes (this will also pull steps)
                    let pulled = try await compositeRepo.pullRoutines(userId: user.id, familyId: user.familyId)
                    if pulled > 0 {
                        AppLogger.ui.info("✅ Pulled \(pulled) routine(s) from Firestore before loading dashboard")
                    }
                } catch {
                    AppLogger.ui.warning("⚠️ Failed to sync routines before loading dashboard: \(error.localizedDescription)")
                    // Continue loading local data even if sync fails
                }
            }
            
            // Load routines (exclude deleted) for the user's family
            let allRoutines = try await routineRepository.getAll(familyId: user.familyId, includeDeleted: false)
            routines = allRoutines.sorted { $0.createdAt < $1.createdAt }
            
            // Load children for the user's family
            children = try await childProfileRepository.getAll(familyId: user.familyId)
            
            AppLogger.ui.info("Loaded \(self.routines.count) routines and \(self.children.count) children for family: \(user.familyId)")
        } catch {
            AppLogger.ui.error("Error loading dashboard data: \(error.localizedDescription)")
            errorMessage = error.localizedDescription
        }
        
        isLoading = false
    }
    
    func deleteRoutine(_ routine: Routine) async {
        do {
            try await routineRepository.softDelete(id: routine.id)
            await loadData() // Reload
            AppLogger.ui.info("Deleted routine: \(routine.title)")
        } catch {
            AppLogger.ui.error("Error deleting routine: \(error.localizedDescription)")
            errorMessage = "Failed to delete routine"
        }
    }
    
    func signOut() {
        do {
            try authRepository.signOut()
            AppLogger.ui.info("User signed out")
        } catch {
            AppLogger.ui.error("Error signing out: \(error.localizedDescription)")
            errorMessage = "Failed to sign out"
        }
    }
}

